com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs

```diff
@@ -1051,36 +1144,48 @@ public unsafe void Deserialize(FastBufferReader reader)
                     throw new OverflowException("Could not deserialize SceneObject: Out of buffer space.");
                 }
                 reader.ReadValue(out Header);
-                var readSize = Header.HasParent ? FastBufferWriter.GetWriteSize(ParentObjectId) : 0;
-                readSize += Header.HasTransform ? FastBufferWriter.GetWriteSize(Transform) : 0;
-                readSize += Header.IsReparented ? FastBufferWriter.GetWriteSize(IsLatestParentSet) + (IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0) : 0;
+                var readSize = 0;
+                if (Header.HasParent)
+                {
+                    readSize += FastBufferWriter.GetWriteSize(ParentObjectId);
+                    readSize += FastBufferWriter.GetWriteSize(WorldPositionStays);
+                    readSize += FastBufferWriter.GetWriteSize(IsLatestParentSet);
+                    // We need to read at this point in order to get the IsLatestParentSet value
+                    if (!reader.TryBeginRead(readSize))
+                    {
+                        throw new OverflowException("Could not deserialize SceneObject: Out of buffer space.");
+                    }
+
+                    // Read the initial parenting related properties
+                    reader.ReadValue(out ParentObjectId);
+                    reader.ReadValue(out WorldPositionStays);
+                    reader.ReadValue(out IsLatestParentSet);
+
+                    // Now calculate the remaining bytes to read
+                    readSize = 0;
+                    readSize += IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0;
+                }
+
+                readSize += Header.HasTransform ? FastBufferWriter.GetWriteSize<TransformData>() : 0;
                 readSize += Header.IsSceneObject ? FastBufferWriter.GetWriteSize<int>() : 0;
 
+                // Try to begin reading the remaining bytes
                 if (!reader.TryBeginRead(readSize))
                 {
                     throw new OverflowException("Could not deserialize SceneObject: Out of buffer space.");
                 }
 
-                if (Header.HasParent)
+                if (IsLatestParentSet)
```

Comment:
Not entirely related to this change, but it's just occurred to me that we could make these ulongs into longs and effectively use the sign bit to encode whether they're set or not... read one byte, if it's negative it's unset, if it's positive then seek back one byte and read it as ulong. That is, assuming the actual value of this doesn't ever need that bit...

Could be a useful function in BytePacker... WriteOptionalValue() that writes 8 bytes with a 0 leading bit if set, or 1 byte with a 1 leading bit if unset.

Curious what you'd think of that (not an action item for this PR, just a thought)

Reply:
Actually it also occurs to me that there are several boolean values here...

What about writing them as a bitmask using `writer.EnterBitwiseContext()`?

Reply:
I think we should have a separate PR where we go through all `INetworkMessage` implementations and see if we can reduce the total size for each message type.  Perhaps not in this PR?